### PR TITLE
Regression logging

### PR DESCRIPTION
**Pull request by atluq**
**Checklist**

- [ ] tests added
- [x] commits conform to the [Commit message format](https://github.com/qlik-oss/gopherciser/blob/master/.github/CONTRIBUTING.md#commit)
- [x] documentation updated

**Squash commit message**

Add logs to be used for regression testing

Add configurable functionality which logs subscribed objects to a regression log file.

**Info**

Filters may need fine tuning.

### CODE REVIEW

File: scenario/actionhandler.go [hunk 1]
```diff
@@ -459,9 +461,70 @@ func (act *Action) endAction(sessionState *session.State, actionState *action.St
...
+			hyperCube := obj.HyperCube()
+			listObject := obj.ListObject()
+
+			err := sessionState.LogEntry.LogRegression(
+				// use unique id SESSION.ACTION.OBJ
+				fmt.Sprintf("%d.%d.%s", sessionState.LogEntry.Session.Session, sessionState.Counters.ActionID.Current(), obj.ID),
+				map[string]interface{}{
+					// "hyperCubeDataPages":  obj.HyperCubeDataPages(),
+					// "hyperCubeStackPages": obj.HyperCubeStackPages(),
+					// "hyperPivotPages":     obj.HyperPivotPages(),
```
Comment (DnlLrssn) [inline scenario/actionhandler.go:512]: adding these later or why commented? for many objects this is where the data will actually be

File: scenario/actionhandler.go [hunk 2]
```diff
@@ -459,9 +461,70 @@ func (act *Action) endAction(sessionState *session.State, actionState *action.St
...
+			if genObj, ok := obj.EnigmaObject.(*enigma.GenericObject); ok {
+				genType = genObj.GenericType
+			}
+
+			hyperCube := obj.HyperCube()
+			listObject := obj.ListObject()
+
+			err := sessionState.LogEntry.LogRegression(
+				// use unique id SESSION.ACTION.OBJ
+				fmt.Sprintf("%d.%d.%s", sessionState.LogEntry.Session.Session, sessionState.Counters.ActionID.Current(), obj.ID),
```
Comment (DnlLrssn) [inline scenario/actionhandler.go:508]: might want to get the actionid from log entry too, for the sake of container actions, other wise the end of the container action will have the actionID of the last subaction.

**Review by DnlLrssn — COMMENTED** [review]
Might want to log in tsv format after all so it can easily used in SENSE
first X rows can be skipped in load script so filter row could stay, then headers then rows in tsv:
FILTERS ["- qNum","+ qStateCounts","+ qGrandTotalRow","+ qPivotDataPages","+ qStackedDataPages","+ qNum"]
ID    DATA    META
X      {...}       {...}

File: scenario/actionhandler.go [hunk 3]
```diff
@@ -459,9 +461,70 @@ func (act *Action) endAction(sessionState *session.State, actionState *action.St
 		containerActionEntry = originalActionEntry
 	}
 
-	err := logResult(sessionState, actionState, actionState.Details, containerActionEntry)
+	if err := logResult(sessionState, actionState, actionState.Details, containerActionEntry); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (atluq) [reply scenario/actionhandler.go:465]: great catch!

File: scenario/actionhandler.go (hunk 1 above)
Comment (atluq) [reply scenario/actionhandler.go:512]: thought there could be some redundancy in the data and left this for discussion.

File: scenario/actionhandler.go (hunk 2 above)
Comment (atluq) [reply scenario/actionhandler.go:508]: I guess we should go with only the logEntry actionID then?

File: scenario/actionhandler.go (hunk 1 above)
Comment (DnlLrssn) [reply scenario/actionhandler.go:512]: there will be for some objects yes, for others the data will not be there.

**Comment by atluq** [comment]
> Might want to log in tsv format after all so it can easily used in SENSE
> first X rows can be skipped in load script so filter row could stay, then headers then rows in tsv:
> FILTERS ["- qNum","+ qStateCounts","+ qGrandTotalRow","+ qPivotDataPages","+ qStackedDataPages","+ qNum"]
> ID DATA META
> X {...} {...}

Yes, the existence of filter and readability made me go for no columns. But usability in Sense is probably more important.

File: scenario/actionhandler.go (hunk 2 above)
Comment (DnlLrssn) [reply scenario/actionhandler.go:508]: I think that's safer bet.

**Comment by DnlLrssn** [comment]
> 
> 
> > Might want to log in tsv format after all so it can easily used in SENSE
> > first X rows can be skipped in load script so filter row could stay, then headers then rows in tsv:
> > FILTERS ["- qNum","+ qStateCounts","+ qGrandTotalRow","+ qPivotDataPages","+ qStackedDataPages","+ qNum"]
> > ID DATA META
> > X {...} {...}
> 
> Yes, the existence of filter and readability made me go for no columns. But usability in Sense is probably more important.

yes readibiliy suffers somewhat, but makes it easier with sense. Enables excel/libreoffice and easier to use *nix commands like `cut`.

File: scenario/actionhandler.go (hunk 1 above)
Comment (atluq) [reply scenario/actionhandler.go]: Uncommented all of them

File: scenario/actionhandler.go [hunk 4]
```diff
@@ -461,16 +461,17 @@ func (act *Action) endAction(sessionState *session.State, actionState *action.St
...
-		return err
-	}
+	logResultErr := logResult(sessionState, actionState, actionState.Details, containerActionEntry)
 	sessionState.LogEntry.LogDebugf("%s END", act.Type)
-
-	if err := logObjectRegressionData(sessionState); err != nil {
-		return err
+	logRegressionErr := logObjectRegressionData(sessionState)
+	var errs *multierror.Error
+	if logResultErr != nil {
```
Comment (DnlLrssn) [inline scenario/actionhandler.go:468]: if `errs` is of type multi.Error, which it is here, you shouldn't need to check the sub errors for nil, when appending
from multierror.Append:
````golang
if e != nil {
	err.Errors = append(err.Errors, e)
}
```

File: scenario/actionhandler.go [hunk 5]
```diff
@@ -459,9 +461,67 @@ func (act *Action) endAction(sessionState *session.State, actionState *action.St
...
 	sessionState.LogEntry.LogDebugf("%s END", act.Type)
-	return errors.WithStack(err)
+	logRegressionErr := logObjectRegressionData(sessionState)
+	var errs *multierror.Error
+	if logResultErr != nil {
+		errs = multierror.Append(errs, logResultErr)
+	}
+	if logRegressionErr != nil {
+		errs = multierror.Append(errs, logRegressionErr)
+	}
```
Comment (DnlLrssn) [inline scenario/actionhandler.go:473]:
```suggestion
		var errs *multierror.Error
		multierror.Append(errs, logResult(sessionState, actionState, actionState.Details, containerActionEntry))
		multierror.Append(logObjectRegressionData(sessionState))
```

**Comment by DnlLrssn** [comment]
suggestion marking multiple lines didn't work very well when there was a comment in the middle of the lines I marked. So logResult is double now

**Comment by atluq** [comment]
now that TestPanicAndFailRecover test fails again and I am unsure why, see `c69bfa1` build.

File: enigmahandlers/objectsmap.go [hunk 6]
```diff
@@ -90,6 +90,25 @@ func (o *ObjectsMap) AddObject(obj *Object) error {
...
+	// 	fmt.Fprintf(os.Stderr, "genobj layout:\n%s\n", string(rawLayout))
+
+	// 	rawProps, err := typedObj.GetPropertiesRaw(context.TODO())
+	// 	if err != nil {
+	// 		panic(err)
+	// 	}
+	// 	fmt.Fprintf(os.Stderr, "genobj props:\n%s\n", string(rawProps))
+	// }
+	// fmt.Fprintln(os.Stderr, "-----------------------------")
+
```
Comment (DnlLrssn) [inline enigmahandlers/objectsmap.go:111]: removed commented code

File: logger/regression.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	}
+)
+
+var filters = marshalFilters(
+	"+qStateCounts",
+	"+qGrandTotalRow",
+	"+qDataPages",
+	"+qPivotDataPages",
+	"+qStackedDataPages",
+	"-qNum",
```
Comment (DnlLrssn) [inline logger/regression.go:43]: So this these seems to be filters for _hypercube_ top level, so at least "qListObject" for _listObject_ top level should be added.  Should the rest "top level" data also be in filter?
"hyperCubeDataPages"
"hyperCubeStackPages"
"hyperCubePivotPages"
"listObjectDataPages"
as all data in these should be considered significant.

File: scenario/actionhandler.go [hunk 8]
```diff
@@ -459,8 +461,63 @@ func (act *Action) endAction(sessionState *session.State, actionState *action.St
...
+					"hyperCube":           obj.HyperCube(),
+					"listObject":          obj.ListObject(),
+					"listObjectDataPages": obj.ListObjectDataPages(),
+				},
+				map[string]interface{}{
+					"actionType":  sessionState.LogEntry.Action.Action,
+					"actionLabel": sessionState.LogEntry.Action.Label,
+					"actionID":    sessionState.LogEntry.Action.ActionID,
+					"objectID":    obj.ID,
+					"objectType":  genObj.Type,
```
Comment (DnlLrssn) [inline scenario/actionhandler.go:515]: genObj.Type will almost always be GenericObject, genObj.GenericType might be more useful.

File: logger/regression.go (hunk 7 above)
Comment (SeBBBe) [reply logger/regression.go:43]: The filters above match what we have for the SDKExerciser regression logging. Not saying that means it's perfectly correct. We might have to play around with the filters once we have the actual differ to see that they have the intended results.

File: logger/regression.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+// NewRegressionLogger creates a new RegressionLoggerCloser with headerEntries
+// written in the header of the log.
+func NewRegressionLogger(w io.WriteCloser, headerEntries ...HeaderEntry) RegressionLoggerCloser {
+	fmt.Fprintf(w, "HEADER_KEY\tHEADER_VALUE\n")
+	fmt.Fprintf(w, "FILTERS\t%s\n", filters)
+	for _, he := range headerEntries {
+		fmt.Fprintf(w, "%s\t%s\n", strings.ToUpper(strings.TrimSpace(he.Key)), strings.TrimSpace(he.Value))
+	}
+	fmt.Fprintln(w, "---")
+	fmt.Fprintln(w, "ID\tMETA\tDATA")
```
Comment (SeBBBe) [inline logger/regression.go:63]: So we'll have a tsv with different amount of columns depending on row position? Is this format digestible by Sense etc?

**Review by SeBBBe — COMMENTED** [review]
Looks pretty good!

File: scenario/actionhandler.go (hunk 8 above)
Comment (atluq) [reply scenario/actionhandler.go:515]: Yes, agree. Tried both before and dont know why this one ended up in the metadata.

File: logger/regression.go (hunk 9 above)
Comment (atluq) [reply logger/regression.go:63]: Well, if we could start reading from after `---`. I have not tried yet.

File: logger/regression.go (hunk 7 above)
Comment (DnlLrssn) [reply logger/regression.go:43]: The data the filter in SDKExerciser is working on looks different, and my comments wouldn't apply there, except for the qListObject one.

Comment (atluq) [reply logger/regression.go:43]: If we know data under these top level keys shall be fully included, I can add this now. 

Then we can fine tune as we have the diff tool.

File: logger/regression.go (hunk 9 above)
Comment (DnlLrssn) [reply logger/regression.go:63]: you can set the load script in sense to skip the first X rows

File: logger/regression.go (hunk 7 above)
Comment (DnlLrssn) [reply logger/regression.go:43]: for many objects the actual data will be in e.g. _hyperCubeDataPages_ not _hypercube_ using this structure, this is different from how it works in SDKExerciser which works directly on traffic

Comment (DnlLrssn) [reply logger/regression.go:43]: only objects which gets their data in _GetLayout_ request will have the significant portion of data in _hypercube_. This is a large chunk of the objects though.

File: logger/regression.go (hunk 9 above)
Comment (atluq) [reply logger/regression.go:63]: Okay. Then we would have to fix the number of rows in the first part, or chose number of rows dynamically searching for "---". 

My idea was for us to be able to add new metadata rows in the first part (like `ID_FORMAT`).

File: logger/regression.go (hunk 7 above)
Comment (atluq) [reply logger/regression.go:43]: Okay, we still need both 👍 

And `hyperCubeDataPages/**` shall all be included?

File: logger/regression.go (hunk 9 above)
Comment (DnlLrssn) [reply logger/regression.go:63]: yea fixed would be better from sense perspective

File: logger/regression.go (hunk 7 above)
Comment (atluq) [reply logger/regression.go:43]: Added some unit tests where we can test the intended behavior for real logs.

**Review by DnlLrssn — APPROVED** [review]
Other than the filters which we can update at a later point

**Comment by DnlLrssn** [comment]
@atluq please hold of merging this until next release has been done, there might be more bugfixes incoming before release is set.

**Comment by atluq** [comment]
> @atluq please hold of merging this until next release has been done, there might be more bugfixes incoming before release is set.

Yes. Will not merge until the full pipe of regression analysis is ready. It unnecessary to release this until then.

**Review by SeBBBe — APPROVED** [review]

**Comment by SeBBBe** [comment]
Selecting schedulers is not fully supported in the GUI, just FYI. We can probably get it working but it will require some manual effort.

**Comment by atluq** [comment]
> Selecting schedulers is not fully supported in the GUI, just FYI. We can probably get it working but it will require some manual effort.

Great to know. This commit disables forced scheduler and automatically overrides any scheduler with a preconfigured simpleScheduler, both only if regression logging is enabled. So no manual scheduler selection is needed.

**Comment by SeBBBe** [comment]
> > Selecting schedulers is not fully supported in the GUI, just FYI. We can probably get it working but it will require some manual effort.
> 
> Great to know. This commit disables forced scheduler and automatically overrides any scheduler with a preconfigured simpleScheduler, both only if regression logging is enabled. So no manual scheduler selection is needed.

Ah, right. Didn't look at it carefully enough I guess. That's a good way of doing it - we should add logic in the frontend to hide the irrelevant settings when the checkbox is ticked.

**Comment by atluq** [comment]
> we should add logic in the frontend to hide the irrelevant settings when the checkbox is ticked.

Good point, will think of a good way to do that 👍

File: scheduler/regression.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+			Scheduler: Scheduler{
+				SchedType:      SchedSimple,
+				InstanceNumber: 0,
+			},
+			Settings: SimpleSchedSettings{
+				ExecutionTime:   -1,
+				Iterations:      1,
+				ConcurrentUsers: 1,
+				RampupDelay:     0.5,
+			},
```
Comment (DnlLrssn) [inline scheduler/regression.go:17]: init SimpleScheduler as 

```golang
&SimpleScheduler{Scheduler{SchedSimple,0,}, SimpleSchedSettings{.etc}
```

And we would get a compile error if we add an entry = not forget to add that here.

**Review by SeBBBe — APPROVED** [review]
👍

File: scheduler/regression.go (hunk 10 above)
Comment (atluq) [reply scheduler/regression.go]: Changing them all reads pretty bad 🤔
```
			Scheduler{
				SchedSimple,
				TimeBuffer{},
				0,
				session.ReconnectSettings{},
				nil,
				false,
			},
			SimpleSchedSettings{
				-1,
				1,
				0.5,
				1,
				false,
				false,
			},
```

**Pull request merged**